README
                 
This is fork of Routino by Twinslah. Website: http://twinslash.com

=========================================================================

		  Routino : OpenStreetMap Routing Software
                  ========================================


   Routino is an application for finding a route between two points using
   the dataset of topographical information collected by
   http://www.OpenStreetMap.org.

   Starting from the raw OpenStreetMap data (in the form of the '.osm' XML
   files available on the internet) a custom database is generated that
   contains the information useful for routing. With this database and two
   points specified by latitude and longitude an optimum route (either
   shortest or quickest) is determined. The route is calculated for
   OpenStreetMap highways (roads, paths etc) using one of the common forms
   of transport defined in OpenStreetMap (foot, bicycle, horse, motorcar,
   motorbike etc).

   When processing the OpenStreetMap data the types of highways are
   recorded and these set default limits on the types of traffic allowed.
   More specific information about permissions for different types of
   transport are also recorded as are maximum speed limits. Further
   restrictions like oneway streets, weight, height, width and length
   limits are also included where specified. Additionally a set of
   properties of each highway are also recorded. The processing of the
   input file is controlled by a configuration file which determines the
   information that is used.

   When calculating a route the type of transport to be used is taken into
   account to ensure that the known restrictions are followed. Each of the
   different highway types can further be allowed or disallowed depending
   on preferences. For each type of highway a default speed limit is
   defined (although the actual speed used will be the lowest of the
   default and any specified in the original data). To make use of the
   information about restrictions the weight, height, width and length of
   the transport can also be specified. Further preferences about road
   properties (e.g. paved or not) can also be selected.

   The result of calculating the route can be presented in several
   different ways. An HTML file can be produced that contains a
   description of the route to take with instructions for each of the
   important junctions. The contents of the file are created based on a
   set of translations specified in a configuration file. The route is
   also available in a GPX (GPS eXchange) XML format. format file
   containing either every point and highway segment (a track file) or
   just a waypoint and translated instructions for the important junctions
   (a route file). Additionally there are two plain text files that
   contain all data points or just the important ones (intended for
   debugging and further processing).

   One of the design aims of Routino was to make the software are flexible
   as possible in selecting routing preferences but also have a sensible
   set of default values. Another design aim was that finding the optimum
   route should be very fast and most of the speed increases come from the
   carefully chosen and optimised data format.


Disclaimer
----------

   The route that is calculated by this software is only as good as the
   input data.

   Routino comes with ABSOLUTELY NO WARRANTY for the software itself or
   the route that is calculated by it.


Demonstration
-------------

   A live demonstration of the router for the UK is available on the
   internet:

   http://www.routino.org/uk/router.html

   The source code download available also includes a set of files that can
   be used to create your own interactive map.

   The interactive map is made possible by use of the OpenLayers Javascript
   library from http://www.openlayers.org/.


Documentation
-------------

   The algorithm used is described in the file ALGORITHM.txt and some
   notes about the limitations of the data is in DATA.txt.

   The configuration files and in particular the default set of rules for
   processing the OpenStreetMap data tags are described in detail in
   CONFIGURATION.txt and TAGGING.txt.  The format of the output files
   generated are described in OUTPUT.txt.

   Detailed information about how to use the programs is available in the
   file USAGE.txt and how to install it is in INSTALL.txt.


Status
------

   Version 1.0 of Routino was released on 8th April 2009.
   Version 1.1 of Routino was released on 13th June 2009.
   Version 1.2 of Routino was released on 21st October 2009.
   Version 1.3 of Routino was released on 21st January 2010.
   Version 1.4 of Routino was released on 31st May 2010.


License
-------

   This program is free software: you can redistribute it and/or modify it
   under the terms of the GNU Affero General Public License as published
   by the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   It is important to note that for this program I have decided to use the
   Affero GPLv3 instead of just using the GPL. This license adds
   additional requirements to anybody who provides a networked service
   using this software.


Copyright
---------

   Routino is copyright Andrew M. Bishop 2008-2010.

   Contact amb@gedanken.demon.co.uk for any questions or queries.


Homepage
--------

   The latest information about the program can be found on the homepage:

   http://www.routino.org/


Download
--------

   The program can be downloaded from:

   http://www.routino.org/download/


--------

Copyright 2008-2010 Andrew M. Bishop.


ChangeLog
2010-07-10  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	Version 1.4.1 released

2010-07-10  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* doc/NEWS.txt: Update NEWS for release.

	* doc/ALGORITHM.txt:
	Update documentation for slight modification to algorithm, also add more
	information about how preferences etc are handled.

2010-07-09  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/Makefile:
	Default compilation flags include optimisation and not debugging symbols.

2010-07-08  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/nodes.c:
	Fix error with finding closest segment to the specified point.

	* src/optimiser.c:
	Bug fix for not crashing when finding the middle part of the route.

2010-07-07  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/results.c, src/optimiser.c:
	Changed the amount of memory allocated for intermediate results => routes much
	faster.

	* src/output.c: Remove compilation warning.

	* src/Makefile:
	Copy files to web directory like done in other Makefiles.

	* doc/Makefile:
	Change location of HTML files in web directory and clean up web directory on distclean.

	* src/xml/Makefile: Stop message being printed when make runs.

	* xml/Makefile:
	Fix error from last checkin and clean up web directory on distclean.

2010-07-06  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/optimiser.c:
	Don't crash if the middle part of the route can't be found but exit cleanly.

2010-07-05  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/superx.c: Change the algorithm used to determine supernodes.

2010-07-03  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xml/routino-translations.xml:
	Added German translation [patch from Christoph Eckert].

	* src/translations.c:
	Don't crash if more than one language is in translations.xml but --language
	option is not used.

2010-06-28  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/router.c: Don't crash if start and finish are the same point.

2010-06-27  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* doc/DATA.txt: New file.

	* doc/ALGORITHM.txt, doc/CONFIGURATION.txt, doc/INSTALL.txt, doc/OUTPUT.txt, doc/README.txt,
	doc/TAGGING.txt, doc/USAGE.txt:
	Updated documentation to match new web site.

	* doc/Makefile: New file.

	* xml/Makefile: Add some new variables.

2010-06-26  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xml/routino-profiles.xml, xml/routino-tagging-nomodify.xml, xml/routino-tagging.xml,
	xml/routino-translations.xml, src/translations.c:
	Changed URLs to http://www.routino.org/

	* doc/README.txt: *** empty log message ***

	* doc/OUTPUT.txt: Changed URLs to http://www.routino.org/

2010-05-31  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	Version 1.4 released

2010-05-31  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* doc/INSTALL.txt, doc/NEWS.txt, doc/README.txt:
	Update for version 1.4.

	* src/xml/Makefile: Make sure that distclean really cleans up.

	* Makefile: Make sure that xml sub-directory is made.

	* src/router.c:
	Fix the code that should stop routing if no segment is found.

2010-05-30  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* doc/USAGE.txt:
	Add the planetsplitter tagging rules option (and remove the unnecessary options
	that it replaces), add the filedumper OSM dump option and add the tagmodifier
	program.

	* doc/TAGGING.txt: Describe the new tagging rules.

	* doc/OUTPUT.txt: Note that the HTML and GPX outputs are translated.

	* doc/CONFIGURATION.txt: Add the tagging rules configuration file.

	* doc/ALGORITHM.txt: An update to the current size of the UK database.

	* xml/routino-tagging-nomodify.xml: New file.

	* src/tagmodifier.c:
	A tagging XML file must be read (just like planetsplitter).

	* src/filedumper.c:
	Add the option to dump a region rather than all and to not output super
	segments.

	* src/optimiser.c: Fix printing the number of super-segments tried.

2010-05-29  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xml/routino-translations.xml, xml/routino-translations.xsd, src/ways.h, src/filedumper.c,
	src/osmparser.c, src/output.c, src/translations.c, src/translations.h:
	Translate the names given to unnamed roads (the highway type).

	* src/profiles.c, src/profiles.h, src/router.c:
	Stricter check on specified profile before routing.

	* src/router.c: Ensure that if no segment is found the routing stops.

	* src/nodes.c:
	When finding a closest segment one of the nodes must be within the search
	distance.

2010-05-28  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/router.c: Make sure that some profiles are loaded.

2010-05-27  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/optimiser.c, src/profiles.c:
	Fix bug with profile preferences (used incorrectly in route optimisation).

	* src/Makefile, src/filedumper.c, src/types.c, src/types.h:
	Add an option to filedumper to dump an OSM format file.

2010-05-25  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/xmlparse.l: Fix bug with encoding XML strings.

2010-05-23  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* xml/Makefile:
	Make sure that modified files are copied to web directory.

	* src/tagmodifier.c:
	Fix bug when filename is specified on command line.

	* src/tagging.c, src/tagging.h, src/tagmodifier.c, src/xmlparse.l, src/osmparser.c:
	Fix some memory leaks.

	* src/tagmodifier.c, xml/osm.xsd, xml/routino-osm.xsd, src/osmparser.c:
	Add the 'bound' element to the XML parser.

2010-05-22  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/functionsx.h, src/osmparser.c, src/planetsplitter.c, src/ways.h, src/waysx.c,
	src/waysx.h:
	Remove the --transport=<transport>, --not-highway=<highway> and
	--not-property=<property> options from planetsplitter because they can be done
	by the tagging.xml file now.

2010-05-18  Andrew M. Bishop  <amb@gedanken.demon.co.uk>

	* src/Makefile: Add tagmodifier program.

	* src/xmlparse.l: Handle floating point numbers in scientific notation.

	* src/planetsplitter.c:

src/types.c
/***************************************
 $Header: /home/amb/routino/src/RCS/types.c,v 1.3 2010/05/27 17:25:23 amb Exp $

 Functions for handling the data types.

 Part of the Routino routing software.
 ******************/ /******************
 This file Copyright 2008-2010 Andrew M. Bishop

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU Affero General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU Affero General Public License for more details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ***************************************/


#include <string.h>

#include "types.h"


/*++++++++++++++++++++++++++++++++++++++
  Decide on the type of a way given the "highway" parameter.

  Highway HighwayType Returns the highway type of the way.

  const char *highway The string containing the type of the way.
  ++++++++++++++++++++++++++++++++++++++*/

Highway HighwayType(const char *highway)
{
 switch(*highway)
   {
   case 'c':
    if(!strcmp(highway,"cycleway")) return(Way_Cycleway);
    return(Way_Count);

   case 'm':
    if(!strcmp(highway,"motorway")) return(Way_Motorway);
    return(Way_Count);

   case 'p':
    if(!strcmp(highway,"primary")) return(Way_Primary);
    if(!strcmp(highway,"path")) return(Way_Path);
    return(Way_Count);

   case 'r':
    if(!strcmp(highway,"residential")) return(Way_Residential);
    return(Way_Count);

   case 's':
    if(!strcmp(highway,"secondary")) return(Way_Secondary);
    if(!strcmp(highway,"service")) return(Way_Service);
    if(!strcmp(highway,"steps")) return(Way_Steps);
    return(Way_Count);

   case 't':
    if(!strcmp(highway,"trunk")) return(Way_Trunk);
    if(!strcmp(highway,"tertiary")) return(Way_Tertiary);
    if(!strcmp(highway,"track")) return(Way_Track);
    return(Way_Count);

   case 'u':
    if(!strcmp(highway,"unclassified")) return(Way_Unclassified);
    return(Way_Count);

   default:
    ;
   }

 return(Way_Count);
}


/*++++++++++++++++++++++++++++++++++++++
  Decide on the type of transport given the name of it.

  Transport TransportType Returns the type of the transport.

  const char *transport The string containing the method of transport.
  ++++++++++++++++++++++++++++++++++++++*/

Transport TransportType(const char *transport)
{
 switch(*transport)
   {
   case 'b':
    if(!strcmp(transport,"bicycle"))
       return(Transport_Bicycle);
    break;

   case 'f':
    if(!strcmp(transport,"foot"))
       return(Transport_Foot);
    break;

   case 'g':
    if(!strcmp(transport,"goods"))
       return(Transport_Goods);
    break;

   case 'h':
    if(!strcmp(transport,"horse"))
       return(Transport_Horse);
    if(!strcmp(transport,"hgv"))
       return(Transport_HGV);
    break;

   case 'm':
    if(!strcmp(transport,"moped"))
       return(Transport_Moped);
    if(!strcmp(transport,"motorbike"))
       return(Transport_Motorbike);
    if(!strcmp(transport,"motorcar"))
       return(Transport_Motorcar);
    break;

   case 'p':
    if(!strcmp(transport,"psv"))
       return(Transport_PSV);
    break;

   case 'w':
    if(!strcmp(transport,"wheelchair"))
       return(Transport_Wheelchair);
    break;

   default:
    return(Transport_None);
   }

 return(Transport_None);
}


/*++++++++++++++++++++++++++++++++++++++
  Decide on the type of property given the name of it.

  Property PropertyType Returns the type of the property.

  const char *property The string containing the method of property.
  ++++++++++++++++++++++++++++++++++++++*/

Property PropertyType(const char *property)
{
 switch(*property)
   {
   case 'b':
    if(!strcmp(property,"bridge"))
       return(Property_Bridge);
    break;

   case 'm':
    if(!strcmp(property,"multilane"))
       return(Property_Multilane);
    break;

   case 'p':
    if(!strcmp(property,"paved"))
       return(Property_Paved);
    break;

   case 't':
    if(!strcmp(property,"tunnel"))
       return(Property_Tunnel);
    break;

   default:
    return(Property_None);
   }

 return(Property_None);
}


/*++++++++++++++++++++++++++++++++++++++
  A string containing the name of a type of highway.

  const char *HighwayName Returns the name.

  Highway highway The highway type.
  ++++++++++++++++++++++++++++++++++++++*/

const char *HighwayName(Highway highway)
{
 switch(highway)
   {
   case Way_Motorway:
    return("motorway");
   case Way_Trunk:
    return("trunk");
   case Way_Primary:
    return("primary");
   case Way_Secondary:
    return("secondary");
   case Way_Tertiary:
    return("tertiary");
   case Way_Unclassified:
    return("unclassified");
   case Way_Residential:
    return("residential");
   case Way_Service:
    return("service");
   case Way_Track:
    return("track");
   case Way_Cycleway:
    return("cycleway");
   case Way_Path:
    return("path");
   case Way_Steps:
    return("steps");

   case Way_Count:
    ;

   case Way_OneWay:
   case Way_Roundabout:
    ;
   }

 return(NULL);
}


/*++++++++++++++++++++++++++++++++++++++
  A string containing the name of a type of transport.

  const char *TransportName Returns the name.

  Transport transport The transport type.
  ++++++++++++++++++++++++++++++++++++++*/

const char *TransportName(Transport transport)
{
 switch(transport)
   {
   case Transport_None:
    return("NONE");

   case Transport_Foot:
    return("foot");
   case Transport_Horse:
    return("horse");
   case Transport_Wheelchair:
    return("wheelchair");
   case Transport_Bicycle:
    return("bicycle");
   case Transport_Moped:
    return("moped");
   case Transport_Motorbike:
    return("motorbike");
   case Transport_Motorcar:
    return("motorcar");
   case Transport_Goods:
    return("goods");
   case Transport_HGV:
    return("hgv");
   case Transport_PSV:
    return("psv");

   case Transport_Count:
    ;
  }

 return(NULL);
}


/*++++++++++++++++++++++++++++++++++++++
  A string containing the name of a highway property.

  const char *PropertyName Returns the name.

  Property property The property type.
  ++++++++++++++++++++++++++++++++++++++*/

const char *PropertyName(Property property)
{
 switch(property)
   {
   case Property_None:
    return("NONE");

   case Property_Paved:
    return("paved");

   case Property_Multilane:
    return("multilane");

   case Property_Bridge:
    return("bridge");

   case Property_Tunnel:
    return("tunnel");

   case Property_Count:
    ;
  }

 return(NULL);
}


/*++++++++++++++++++++++++++++++++++++++
  A string containing the names of allowed transports on a way.

  const char *AllowedNameList Returns the list of names.

  wayallow_t allowed The allowed type.
  ++++++++++++++++++++++++++++++++++++++*/

const char *AllowedNameList(wayallow_t allowed)
{
 static char string[256];

 string[0]=0;

 if(allowed & Allow_Foot)
    strcat(string,"foot");

 if(allowed & Allow_Horse)
   {
    if(*string) strcat(string,", ");
    strcat(string,"horse");
   }

 if(allowed & Allow_Wheelchair)
   {
    if(*string) strcat(string,", ");
    strcat(string,"wheelchair");
   }

 if(allowed & Allow_Bicycle)
   {
    if(*string) strcat(string,", ");
    strcat(string,"bicycle");
   }

 if(allowed & Allow_Moped)
   {
    if(*string) strcat(string,", ");
    strcat(string,"moped");
   }

 if(allowed & Allow_Motorbike)
   {
    if(*string) strcat(string,", ");
    strcat(string,"motorbike");
   }

 if(allowed & Allow_Motorcar)
   {
    if(*string) strcat(string,", ");
    strcat(string,"motorcar");
   }

 if(allowed & Allow_Goods)
   {
    if(*string) strcat(string,", ");
    strcat(string,"goods");
   }

 if(allowed & Allow_HGV)
   {
    if(*string) strcat(string,", ");
    strcat(string,"hgv");
   }

 if(allowed & Allow_PSV)
   {
    if(*string) strcat(string,", ");
    strcat(string,"psv");
   }

 return(string);
}


/*++++++++++++++++++++++++++++++++++++++
  A string containing the names of the properties of a way.

  const char *PropertiesNameList Returns the list of names.

  wayprop_t properties The properties of the way.
  ++++++++++++++++++++++++++++++++++++++*/

const char *PropertiesNameList(wayprop_t properties)
{
 static char string[256];

 string[0]=0;

 if(properties & Properties_Paved)
   {
    if(*string) strcat(string,", ");
    strcat(string,"paved");
   }

 if(properties & Properties_Multilane)
   {
    if(*string) strcat(string,", ");
    strcat(string,"multilane");
   }

 if(properties & Properties_Bridge)
   {
    if(*string) strcat(string,", ");
    strcat(string,"bridge");
   }

 if(properties & Properties_Tunnel)
   {
    if(*string) strcat(string,", ");
    strcat(string,"tunnel");
   }

 return(string);
}


/*++++++++++++++++++++++++++++++++++++++
  Returns a list of all the highway types.

  const char *HighwayList Return a list of all the highway types.
  ++++++++++++++++++++++++++++++++++++++*/

const char *HighwayList(void)
{
 return "    motorway     = Motorway\n"
        "    trunk        = Trunk\n"
        "    primary      = Primary\n"
        "    secondary    = Secondary\n"
        "    tertiary     = Tertiary\n"
        "    unclassified = Unclassified\n"
        "    residential  = Residential\n"
        "    service      = Service\n"
        "    track        = Track\n"
        "    cycleway     = Cycleway\n"
        "    path         = Path\n"
        "    steps        = Steps\n"
        ;
}


/*++++++++++++++++++++++++++++++++++++++
  Returns a list of all the transport types.

  const char *TransportList Return a list of all the transport types.
  ++++++++++++++++++++++++++++++++++++++*/

const char *TransportList(void)
{
 return "    foot       = Foot\n"
        "    bicycle    = Bicycle\n"
        "    wheelchair = Wheelchair\n"
        "    horse      = Horse\n"
        "    moped      = Moped     (Small motorbike, limited speed)\n"
        "    motorbike  = Motorbike\n"
        "    motorcar   = Motorcar\n"
        "    goods      = Goods     (Small lorry, van)\n"
        "    hgv        = HGV       (Heavy Goods Vehicle - large lorry)\n"
        "    psv        = PSV       (Public Service Vehicle - bus, coach)\n"
        ;
}


/*++++++++++++++++++++++++++++++++++++++
  Returns a list of all the property types.

  const char *PropertyList Return a list of all the highway proprties.
  ++++++++++++++++++++++++++++++++++++++*/

const char *PropertyList(void)
{
 return "    paved     = Paved (suitable for normal wheels)\n"
        "    multilane = Multiple lanes\n"
        "    bridge    = Bridge\n"
        "    Tunnel    = Tunnel\n"
        ;
}
